lib/api/formtools/grid.js

```diff
@@ -0,0 +1,263 @@
+        }
+
+        if (!modelName || typeof(modelName) !== 'string' || !modelName.length) {
+            throw new Error('modelName is required and must be of type string and cannot be emptied.');
+        }
+
+        if (!data || typeof(data) !== 'object') {
+            throw new Error('data is required and must be of type object.');
+        }
+
+        var filters = {},
+            model = linz.get('models')[modelName];
+
+        async.waterfall([
+
+            function (callback) {
+                return getSettings(modelName, callback);
+            }
+
+        ], function (err, settings) {
+
+            if (err) {
+                return cb(err);
+            }
+
+            if (!Object.keys(settings.filters).length) {
+                return cb(null);
+            }
+
+            async.each(fieldNames, function (fieldName, filtersDone) {
+
+                if (!data[fieldName]) {
+                    return filtersDone(null);
+                }
+
+                // call the filter renderer and update the content with the result
+                settings.filters[fieldName].filter.filter(fieldName, data, function (filterErr, result) {
+
+                    if (filterErr) {
+                        return filtersDone(filterErr);
+                    }
+
+                    filters = model.addSearchFilter(filters, result);
+
+                    return filtersDone(null);
+
+                });
+
+            }, function (filterErr) {
+
+                if (filterErr) {
+                    return cb(filterErr);
+                }
+
+                // consolidate filters into query
+                filters = model.setFiltersAsQuery(filters);
+
+                return cb(filterErr, filters);
+            });
+
+        });
+
+    }
+
+    function getActiveFilters (selectedFilters, data, modelName, cb) {
+
+        if (!selectedFilters || !Array.isArray(selectedFilters) || !selectedFilters.length) {
+            throw new Error('selectedFilters is required and must be of type array and cannot be emptied.');
+        }
+
+        if (!modelName || typeof(modelName) !== 'string' || !modelName.length) {
+            throw new Error('modelName is required and must be of type string and cannot be emptied.');
+        }
+
+        if (!data || typeof(data) !== 'object') {
+            throw new Error('data is required and must be of type object.');
+        }
+
+        var model = linz.get('models')[modelName],
+            activeFilters = {};
+
+        async.waterfall([
+
+            function (callback) {
+                return getSettings(modelName, callback);
+            }
+
+        ], function (err, settings) {
+
+            if (err) {
+                return cb(err);
+            }
+
+            if (!Object.keys(settings.filters).length) {
+                return cb(null);
+            }
+
+            async.each(selectedFilters, function (fieldName, filtersDone) {
```

Comment:
This is being used to populate `activeFilters`. Perhaps it would be better to use `async.filter`, chained with `async.map` and then return the result to the callback?